libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+flatpak_remote_ref_actions_file_new (GPtrArray *remote_ref_actions,
+                                     gint       priority)
+{
+  FlatpakRemoteRefActionsFile *file = g_slice_new0 (FlatpakRemoteRefActionsFile);
+
+  file->remote_ref_actions = g_ptr_array_ref (remote_ref_actions);
+  file->priority = priority;
+
+  return file;
+}
+
+void
+flatpak_remote_ref_actions_file_free (FlatpakRemoteRefActionsFile *file)
+{
+  g_clear_pointer (&file->remote_ref_actions, g_ptr_array_unref);
+
+  g_slice_free (FlatpakRemoteRefActionsFile, file);
+}
+
+gboolean
+eos_updater_util_flatpak_ref_actions_append_from_directory (const gchar   *relative_parent_path,
+                                                            GFile         *directory,
+                                                            GHashTable    *ref_actions_for_files,
+                                                            gint           priority,
+                                                            GCancellable  *cancellable,
+                                                            GError       **error)
+{
+  g_autoptr(GFileEnumerator) autoinstall_d_enumerator = NULL;
+
+  /* Repository checked out, read all files in order and build up a list
+   * of flatpaks to auto-install */
+  autoinstall_d_enumerator = g_file_enumerate_children (directory,
+                                                        G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                                        G_FILE_QUERY_INFO_NONE,
+                                                        cancellable,
+                                                        error);
+
+  if (!autoinstall_d_enumerator)
+    return FALSE;
+
+  while (TRUE)
+    {
+      GFile *file;
+      GFileInfo *info;
+      g_autoptr(GPtrArray) action_refs = NULL;
+      const gchar *filename = NULL;
+      g_autoptr(FlatpakRemoteRefActionsFile) actions_file = NULL;
+      FlatpakRemoteRefActionsFile *existing_actions_file = NULL;
+
+      if (!g_file_enumerator_iterate (autoinstall_d_enumerator,
+                                      &info, &file,
+                                      cancellable, error))
+        return FALSE;
+
+      if (!file || !info)
+        break;
+
+      action_refs = read_flatpak_ref_actions_from_file (file,
+                                                        cancellable,
+                                                        error);
+
+      if (!action_refs)
+        return FALSE;
+
+      filename = g_file_info_get_name (info);
+
+      /* We may already have a remote_ref_actions_file in the hash table
+       * and we cannot just blindly replace it. Replace it only if
+       * the priority of the incoming directory is less */
+      existing_actions_file = g_hash_table_lookup (ref_actions_for_files,
+                                                   filename);
+
+      if (existing_actions_file &&
+          existing_actions_file->priority < priority)
+        continue;
```

Comment:
You could do this check before calling `read_flatpak_ref_actions_from_file()` and avoid parsing the file unnecessarily.